File: backend/api/schemas/tryoff.py
import uuid
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class GarmentType(str, Enum):
    upper = "upper"
    lower = "lower"
    dress = "dress"


class TryoffJobStatus(str, Enum):
    pending = "pending"
    processing = "processing"
    complete = "complete"
    failed = "failed"


class TryoffJobRequest(BaseModel):
    source_image_id: uuid.UUID
    garment_type: GarmentType


class TryoffBatchRequest(BaseModel):
    source_image_id: uuid.UUID
    garment_types: list[GarmentType] = Field(..., min_length=1)


class TryoffJobResponse(BaseModel):
    job_id: uuid.UUID
    status: TryoffJobStatus
    garment_type: GarmentType
    created_at: datetime


class TryoffBatchResponse(BaseModel):
    jobs: list[TryoffJobResponse]


class TryoffJobStatusResponse(BaseModel):
    job_id: uuid.UUID
    status: TryoffJobStatus
    garment_type: GarmentType
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    result_url: str | None = None
    error_reason: str | None = None
    retry_count: int = 0


class TryoffJobHistoryItem(BaseModel):
    job_id: uuid.UUID
    status: TryoffJobStatus
    garment_type: GarmentType
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    result_url: str | None = None
    error_reason: str | None = None
    retry_count: int = 0


class TryoffJobHistoryResponse(BaseModel):
    items: list[TryoffJobHistoryItem]
    total: int
    page: int
    page_size: int

File: frontend/lib/api/tryoff.ts
import { apiFetch } from "@/lib/api";

export type GarmentType = "upper" | "lower" | "dress";

export type TryoffJobStatus = "pending" | "queued" | "processing" | "complete" | "failed";

export interface TryoffJob {
  job_id: string;
  garment_type: GarmentType;
  status: TryoffJobStatus;
  source_image_url: string | null;
  source_image_thumb: string | null;
  output_url: string | null;
  output_media_id: string | null;
  error_reason: string | null;
  created_at: string;
  completed_at: string | null;
}

export interface BatchSubmitInput {
  source_image_id: string;
  garment_types: GarmentType[];
}

export interface BatchSubmitResponse {
  jobs: Array<{
    job_id: string;
    status: TryoffJobStatus;
    garment_type: GarmentType;
    created_at: string;
  }>;
}

export interface SourceImageUploadResponse {
  id: string;
  presigned_url: string;
  filename: string;
  uploaded_at: string;
}

export async function uploadTryoffSourceImage(file: File): Promise<SourceImageUploadResponse> {
  const formData = new FormData();
  formData.append("file", file);
  return apiFetch("/api/tryoff/source-images", {
    method: "POST",
    body: formData,
  });
}

export async function submitBatchJobs(
  input: BatchSubmitInput
): Promise<BatchSubmitResponse> {
  return apiFetch("/api/tryoff/jobs/batch", {
    method: "POST",
    body: JSON.stringify(input),
  });
}

export async function getTryoffJob(jobId: string): Promise<TryoffJob> {
  return apiFetch(`/api/tryoff/jobs/${jobId}`);
}

interface TryoffJobStatusApi {
  job_id: string;
  status: TryoffJobStatus;
  garment_type: GarmentType;
  created_at: string;
  completed_at: string | null;
  result_url: string | null;
  output_media_id: string | null;
  error_reason: string | null;
}

function mapTryoffJob(raw: TryoffJobStatusApi): TryoffJob {
  return {
    job_id: raw.job_id,
    garment_type: raw.garment_type,
    status: raw.status,
    source_image_url: null,
    source_image_thumb: null,
    output_url: raw.result_url,
    output_media_id: raw.output_media_id,
    error_reason: raw.error_reason,
    created_at: raw.created_at,
    completed_at: raw.completed_at,
  };
}

export async function getTryoffJobs(jobIds: string[]): Promise<TryoffJob[]> {
  const ids = jobIds.join(",");
  const data = (await apiFetch(
    `/api/tryoff/jobs?job_ids=${encodeURIComponent(ids)}`
  )) as TryoffJobStatusApi[];
  return data.map(mapTryoffJob);
}
